lib/Dialect/SV/Transforms/HWExportModuleHierarchy.cpp

```diff
@@ -0,0 +1,135 @@
+static void printHierarchy(hw::InstanceOp &inst, SymbolTable &symbolTable,
+                           llvm::raw_ostream &os) {
+  os << "{"
+     << "\"instance_name\":\"" << inst.instanceName() << "\","
+     << "\"module_name\":\"" << inst.moduleName() << "\","
+     << "\"instances\":[";
+
+  auto moduleOp =
+      symbolTable.lookup<hw::HWModuleOp>(inst.moduleNameAttr().getValue());
+
+  // Only recurse on module ops, not extern or generated ops, whose internal are
+  // opaque.
+  if (moduleOp) {
+    bool firstIteration = true;
+    for (auto op : moduleOp.getOps<hw::InstanceOp>()) {
+      if (firstIteration)
+        firstIteration = false;
+      else
+        os << ",";
+      printHierarchy(op, symbolTable, os);
+    }
+  }
+
+  os << "]}";
+}
+
+/// Return the JSON-serialized module hierarchy for the given module as the top.
+static std::string extractHierarchyFromTop(hw::HWModuleOp op,
+                                           SymbolTable &symbolTable) {
+  std::string resultBuffer;
+  llvm::raw_string_ostream os(resultBuffer);
+
+  // As a special case for top-level module, set instance name to module name,
+  // since the top-level module is not instantiated.
+  os << "{"
+     << "\"instance_name\":\"" << op.getName() << "\","
+     << "\"module_name\":\"" << op.getName() << "\","
+     << "\"instances\":[";
+
+  bool firstIteration = true;
+  for (auto op : op.getOps<hw::InstanceOp>()) {
+    if (firstIteration)
+      firstIteration = false;
+    else
+      os << ",";
+    printHierarchy(op, symbolTable, os);
+  }
+  os << "]}";
+
+  return resultBuffer;
+}
+
+/// Find the modules corresponding to the firrtl mainModule and DesignUnderTest,
+/// and if they exist, emit a verbatim op with the module hierarchy for each.
+void HWExportModuleHierarchyPass::runOnOperation() {
+  mlir::ModuleOp mlirModule = getOperation();
+  auto builder = OpBuilder::atBlockEnd(mlirModule.getBody());
+  SymbolTable symbolTable(mlirModule);
+
+  if (auto mainModuleName =
+          mlirModule->getAttrOfType<mlir::StringAttr>("firrtl.mainModule")) {
+    auto mainModule = symbolTable.lookup<hw::HWModuleOp>(mainModuleName);
+    assert(mainModule &&
+           "firrtl.mainModule should match the name of some HWModuleOp");
+
+    auto verbatimOp = builder.create<sv::VerbatimOp>(
+        builder.getUnknownLoc(),
+        extractHierarchyFromTop(mainModule, symbolTable));
+    verbatimOp->setAttr(
+        "output_file", hw::OutputFileAttr::get(
+                           builder.getStringAttr(""),
+                           builder.getStringAttr("testharness_hier.json"),
+                           /*exclude_from_filelist=*/builder.getBoolAttr(true),
+                           /*exclude_replicated_ops=*/builder.getBoolAttr(true),
+                           &getContext()));
+  }
+
+  for (auto op : mlirModule.getOps<hw::HWModuleOp>()) {
+    if (op->hasAttr("firrtl.DesignUnderTest")) {
```

Comment:
we should lower this annotation to module level attribute so you don't have to search trough the modules to find the module.

Ideally, from the firrtl annotations, we should lower to a generic attribute, either a "dump from here to file x" (hwmodule attr) or "dump @foo to file x" (module attr)

Reply:
Yeah, thanks, this makes a lot of sense and would simplify other things. For the generic attribute's name, should I still prefix it with `firrtl`? I am fearful of using the global namespace because of the reasoning in https://github.com/llvm/circt/pull/1785#discussion_r708479118 :)

Reply:
I've gone back and forth on it.  If it was in the IR, I'd make it a well-defined, non-firrtl thing (though that definition would have to account for recursive hierarchies).  But this probably should move out of the compiler completely at some point, in which case the firrtl namespace is less likely to acquire dependencies.